[ML] AIOps: Adds Pattern analysis tab in Discover

jgowdyelastic opened the pull request:
Closes https://github.com/elastic/kibana/issues/178534

Replaces the pattern analysis flyout in Discover with a tab which sits alongside the Documents and Field statistics tabs.

![image](https://github.com/elastic/kibana/assets/22172091/027f7751-c61e-4b7e-9625-dd876730ff2e)



**Field selection**
Lists all of the text fields in the index. Auto selects `message`, then `error.message`, then `event.original` and if none of these fields are available, it just selects the first field in the list.

![image](https://github.com/elastic/kibana/assets/22172091/6ee0eb75-ed13-4c16-beb6-3de357dc182c)



The Options menu provides some configuration options:

**Minimum time range**
Sets the minimum time range used for the pattern analysis search. The pattern matching results results will be more accurate the more data it sees, so if the user has selected e.g. last 15mins in the time picker, this settings will ensure a wider time range is used to improve the accuracy of the patterns. If the time picker has a larger time range than this setting, the larger time range will be used.

**Random sampling**
Improves the search performance by using a random sampler. This is the same setting as before.

![image](https://github.com/elastic/kibana/assets/22172091/7a2580f6-61f7-4053-9ac1-93a8e8e2f01c)

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

jgowdyelastic commented:
/ci

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

peteharverson commented on x-pack/test/functional/apps/aiops/log_pattern_analysis_in_discover.ts:
```diff
@@ -23,6 +22,7 @@ export default function ({ getPageObjects, getService }: FtrProviderContext) {
     });
   }
 
+  // temporarily disabled, to be replaced
```
Is this comment still applicable?

jgowdyelastic replied:
Removed in https://github.com/elastic/kibana/pull/178916/commits/d5ed385c8fafaec9a24149e96491d154787e01a7

peteharverson commented on x-pack/plugins/aiops/public/components/log_categorization/log_categorization_for_embeddable/embeddable_menu.tsx:
```diff
@@ -0,0 +1,157 @@
+          label={i18n.translate(
+            'xpack.aiops.logCategorization.embeddableMenu.minimumTimeRangeOptionsRowLabel',
+            {
+              defaultMessage: 'Minimum time range',
```
This field could do with an info tooltip. Challenge will be how to simply explain what it does :)

peteharverson replied:
Added tooltip LGTM

peteharverson commented on x-pack/plugins/aiops/public/components/log_categorization/sampling_menu/random_sampler.ts:
```diff
@@ -28,27 +28,33 @@ export type RandomSamplerProbability = number | null;
+    inputDisplay: i18n.translate(
+      'xpack.aiops.logCategorization.randomSamplerPreference.onAutomaticLabel',
+      {
+        defaultMessage: 'On - automatic',
```
Guess it isn't related to the changes here, but sometimes can see too high precision for the probability:

<img width="442" alt="Screenshot 2024-05-14 at 17 29 15" src="https://github.com/elastic/kibana/assets/7405507/ad282167-b5fd-460f-8be1-ec19f85c5f9a">

jgowdyelastic replied:
Fixed in https://github.com/elastic/kibana/pull/178916/commits/3ea27fc97da1f99cf63ed5168f19e1b57cf9f8c1

peteharverson commented on src/plugins/discover/public/application/main/components/layout/discover_main_content.tsx:
```diff
@@ -145,7 +151,16 @@ export const DiscoverMainContent = ({
             </>
-          )}
+          ) : null}
+          {viewMode === VIEW_MODE.PATTERN_LEVEL ? (
```
Why is the data reloaded and rerendered when the doc chart is shown / hidden?

jgowdyelastic replied:
It causes an outer component rerender. If you're in the document or field stats tabs, they also reload.

peteharverson commented on src/plugins/discover/public/application/main/components/layout/discover_main_content.tsx:
```diff
@@ -145,7 +151,16 @@ export const DiscoverMainContent = ({
+            <PatternAnalysisTab
+              dataView={dataView}
+              stateContainer={stateContainer}
+              onAddFilter={() => setDiscoverViewMode(VIEW_MODE.DOCUMENT_LEVEL)}
```
Are there any checks we can add here to prevent this happening if you select a lot of patterns at once to filter in / out (here I tried to filter out 50 patterns in one go):

<img width="662" alt="Screenshot 2024-05-15 at 12 30 52" src="https://github.com/elastic/kibana/assets/7405507/9edc6dfc-1574-4190-b66f-4f528a0291c5">

jgowdyelastic replied:
We could try to predict the size of the url as the user is selecting patterns. However, this will not stop subsequent non-pattern related filters or URL stored settings from tipping it over the limit.
It'd just mean that as the user is selecting patterns we could show a warning that it might be over the limit.

The number of patterns which cause a problem depends on the size of their pattern keys. The user may have selected only a few (50) patterns but the keys are all very long and so might cause a problem compared to a pile of patterns with short keys.

peteharverson commented on x-pack/plugins/aiops/public/components/log_categorization/category_table/category_table.tsx:
```diff
@@ -302,49 +238,49 @@ export const CategoryTable: FC<Props> = ({
+    <EuiInMemoryTable<Category>
+      compressed
+      items={categories}
+      columns={columns}
```
What do you think about including the name of the selected categorization field in the column heading e.g. `error.message examples`? Otherwise the only way to know which field is being used is by opening the Options popover. Would also be helpful in the warning callout and expanded row:

<img width="583" alt="Screenshot 2024-05-15 at 12 45 03" src="https://github.com/elastic/kibana/assets/7405507/acadc081-be91-4d27-9d52-92c26ac910ca">

jgowdyelastic replied:
Updated in https://github.com/elastic/kibana/pull/178916/commits/820b2be0f5ff2884410b273a47eb996f47d6a374
![image](https://github.com/elastic/kibana/assets/22172091/c68b21e1-606d-4ac2-bcc3-34b2266afd57)

jgowdyelastic replied:
I'm thinking now that I should move the field selection out of the menu and have it at the top. Clicking it would open a popover with the field select.
<img width="411" alt="image" src="https://github.com/elastic/kibana/assets/22172091/e2aadf29-4f85-4973-a35b-57e8204dc5f1">

peteharverson replied:
I like the separate control for selecting the field.

peteharverson commented:
Noting that the Discover 'Save' currently saves out the search into the documents mode from the Patterns tab (in contrast to when in the Field Statistics tab that it saves in the Field Statistics mode). If you then switch from Patterns to Documents, the 'Unsaved changes' badge appears in the top nav, which is a bit misleading as in reality the saved search / mode won't have changed.

With the plans to add a log patterns embeddable soon, it's probably not worth spending time trying to address.

peteharverson commented on x-pack/plugins/aiops/public/components/log_categorization/category_table/category_table.tsx:
```diff
@@ -302,49 +238,49 @@ export const CategoryTable: FC<Props> = ({
+      itemId="key"
+      onTableChange={onTableChange}
+      pagination={pagination}
+      sorting={sorting}
```
The direction of sort on the Count column isn't being persisted when the view refreshes.

jgowdyelastic replied:
Fixed in https://github.com/elastic/kibana/pull/178916/commits/c890187dbc565c8ac852acf472e58c57730c2fb9

qn895 commented on src/plugins/discover/public/application/main/components/layout/discover_main_content.tsx:
```diff
@@ -72,6 +72,8 @@ export const DiscoverMainContent = ({
         if (mode === VIEW_MODE.AGGREGATED_LEVEL) {
           trackUiMetric(METRIC_TYPE.CLICK, FIELD_STATISTICS_VIEW_CLICK);
+        } else if (mode === VIEW_MODE.PATTERN_LEVEL) {
+          trackUiMetric(METRIC_TYPE.CLICK, PATTERN_ANALYSIS_LOADED);
```
This should be PATTERN_ANALYSIS_VIEW_CLICK

jgowdyelastic replied:
Updated in https://github.com/elastic/kibana/pull/178916/commits/186f78b53ebc6cf2166bb73a2daffa6f267ef003

qn895 commented on src/plugins/discover/public/application/main/components/layout/discover_main_content.tsx:
```diff
@@ -128,15 +133,16 @@ export const DiscoverMainContent = ({
           {viewMode === VIEW_MODE.DOCUMENT_LEVEL ? (
             <DiscoverDocuments
-              viewModeToggle={viewModeToggle}
+              viewModeToggle={viewModeToggle()}
```
For clarity I think we should rename it to something like viewModeToggle={renderViewModeToggle()}

jgowdyelastic replied:
Updated in https://github.com/elastic/kibana/pull/178916/commits/186f78b53ebc6cf2166bb73a2daffa6f267ef003

jgowdyelastic replied:
> super duper nit
Nah, this needed changing. It was being passed about differently, for the pattern analysis component it was being passed I as a reference, but for the original tabs it was being called and the result passed in.
Definitely needed renaming.

qn895 commented on x-pack/test/functional/apps/ml/data_visualizer/index_data_visualizer_grid_in_discover_trial.ts:
```diff
@@ -0,0 +1,72 @@
+      await esArchiver.loadIfNeeded('x-pack/test/functional/es_archives/ml/farequote');
+      await esArchiver.loadIfNeeded('x-pack/test/functional/es_archives/ml/module_sample_logs');
+      await ml.testResources.createDataViewIfNeeded('ft_farequote', '@timestamp');
+      await ml.testResources.createDataViewIfNeeded('ft_module_sample_logs', '@timestamp');
```
Since this test is only running on `farequoteDataViewTestData`, we can probably remove module_sample_logs and other saved search.

jgowdyelastic replied:
Updated in https://github.com/elastic/kibana/pull/178916/commits/a7ddc2a0776a91b1009bca116a130e114cda23c0

qn895 commented on x-pack/test/functional/apps/ml/data_visualizer/index_data_visualizer_grid_in_discover_trial.ts:
```diff
@@ -0,0 +1,72 @@
+    });
+  }
+
+  describe('field statistics in Discover', function () {
```
We should add the `trial` context for this test suite name.

jgowdyelastic replied:
Updated in https://github.com/elastic/kibana/pull/178916/commits/a7ddc2a0776a91b1009bca116a130e114cda23c0

qn895 commented on x-pack/plugins/aiops/public/components/log_categorization/utils.ts:
```diff
@@ -0,0 +1,60 @@
+  const dataViewFields = dataView.fields.filter((f) => f.esTypes?.includes(ES_FIELD_TYPES.TEXT));
+
+  let messageField: DataViewField | null | undefined = dataViewFields.find(
+    (f) => f.name === 'message'
```
Just to double check here: I'm assuming there's never a case these fields could be capitalized differently? Like 'Message', or 'MESSAGE'. And in those cases, we do we want to support that?

jgowdyelastic replied:
No I don't believe there is a risk of that, these are ECS fields.

jgowdyelastic commented on x-pack/plugins/aiops/public/components/log_categorization/utils.ts:
```diff
@@ -0,0 +1,60 @@
+
+/**
+ * Retrieves the message field from a DataView object.
+ * If the message field is not found, it falls back to error.message or event.original or the first field in the DataView.
```
Updated in https://github.com/elastic/kibana/pull/178916/commits/1fa22ce642fbe0c82b934cd5d510197b978292b9

qn895 commented on src/plugins/discover/public/components/view_mode_toggle/view_mode_toggle.tsx:
```diff
@@ -22,20 +23,60 @@ export const DocumentViewModeToggle = ({
+        .then((patternAnalysisAvailable) => {
+          patternAnalysisAvailable(dataView)
+            .then(setShowPatternAnalysisTab)
+            .catch(() => setShowPatternAnalysisTab(false));
```
Should check if the component is unmounted here or not before setting

qn895 commented:
Tested in observability serverless and I get this warning message. Not sure if it's intentional to have the warning when there are fields shown.
<img width="1917" alt="Screenshot 2024-05-15 at 14 41 06" src="https://github.com/elastic/kibana/assets/43350163/c1a25dce-374c-4858-be75-60c1e2ddfb13">

qn895 commented on x-pack/plugins/aiops/public/components/log_categorization/category_table/category_table.tsx:
```diff
@@ -183,9 +124,15 @@ export const CategoryTable: FC<Props> = ({
+              defaultMessage: 'Examples',
+            })
+          : i18n.translate('xpack.aiops.logCategorization.column.examplesWithFieldName', {
+              defaultMessage: '{fieldName} examples',
```
Wonder if we should add some styling (like a badget?) to the `{fieldName}` here since it was not obvious at first glance that the pattern analysis is being done over 'message' or some other fields. Or Examples for `{fieldName}` to capitalize the first letter in the column name to be consistent.

jgowdyelastic replied:
Yeah, I also don't think it's obvious.
I'll see if a text field icon before the name looks better.
One alternative would be to have something like `Selected field: message` visible somewhere, but there isn't really any room for it, other than bumping the table down a bit and sticking it at the top. But I think that'll leave a lot of empty space on the right.

jgowdyelastic replied:
I've removed this column title in favour of a dedication field selection component. This makes it far more obvious which field has been selected for analysis.
![image](https://github.com/elastic/kibana/assets/22172091/15014e93-a715-4985-bdb0-9b8385214acd)

qn895 commented on x-pack/plugins/aiops/public/components/log_categorization/category_table/table_header.tsx:
```diff
@@ -49,32 +47,57 @@ export const TableHeader: FC<Props> = ({
+  showText = true,
+}) => {
+  const { labels, openFunction } = openInDiscover;
+  const TooltipWrapper: FC<PropsWithChildren<{ text: string }>> = ({ text, children }) => {
```
Should we move `TooltipWrapper` component out and pass showText as a prop so this is not created every time OpenInDiscoverButtons is rendered?

jgowdyelastic replied:
Updated in https://github.com/elastic/kibana/pull/178916/commits/a6b73baba7bab7bb71a1e267c652b1e4253c5ffe

jgowdyelastic commented:
>Tested in observability serverless and I get this warning message. Not sure if it's intentional to have the warning when there are fields shown.

@qn895 That warning looks incorrect, obviously there are values for that field. What dataset is that?
We do need to show a warning if the data isn't good for analysis, even if there are some examples, e.g. for a low token count:
<img width="1002" alt="image" src="https://github.com/elastic/kibana/assets/22172091/ac630b76-eca5-44f6-9743-1e60aff36c6e">

But for the example you've found, that warning is contradicting that fact that there are examples clearly visible.

jughosta reviewed (commented):
Great work, @jgowdyelastic! 

Probably we should introduce a registry for Discover tabs (if they stay around) at some point so you don't have to refactor Discover code when adding new tabs. I think for now it's also fine with direct integration.

There are some minor style differences between Discover UI and Patterns table (spacing, button and text sizes). @andreadelrio would you like to check it? I think it could be improved for visual consistency. 

Overall looking good! Also I left some questions.

This PR would create conflicts with https://github.com/elastic/kibana/pull/180849 PR. Which one are we planning to merge first? cc @qn895

jgowdyelastic commented on x-pack/plugins/aiops/public/components/log_categorization/log_categorization_for_embeddable/discover_tabs.tsx:
```diff
@@ -0,0 +1,115 @@
+          <>
+            <EuiSpacer size="s" />
+            <EuiFlexGroup gutterSize="s" responsive={false}>
+              {selectedCategories.length > 0 ? (
+                <EuiFlexItem>
+                  <SelectedPatterns openInDiscover={openInDiscover} />
+                </EuiFlexItem>
```
The selected count appears when you select patterns using the checkboxes on the left.

szabosteve commented on x-pack/plugins/aiops/public/components/log_categorization/sampling_menu/random_sampler.ts:
```diff
@@ -126,3 +132,58 @@ export class RandomSampler {
+          'xpack.aiops.logCategorization.randomSamplerSettingsPopUp.offCallout.message',
+          {
+            defaultMessage:
+              'Random sampling can be turned on to increase the speed of analysis, although some accuracy will be lost.',
```
```suggestion
              'Random sampling can be turned on to increase analysis speed. Accuracy will slightly decrease.',
```

jgowdyelastic commented on src/plugins/discover/public/application/main/components/layout/discover_main_content.tsx:
```diff
@@ -128,15 +133,16 @@ export const DiscoverMainContent = ({
           {viewMode === VIEW_MODE.DOCUMENT_LEVEL ? (
             <DiscoverDocuments
-              viewModeToggle={viewModeToggle}
+              viewModeToggle={renderViewModeToggle()}
```
Updated in https://github.com/elastic/kibana/pull/178916/commits/299621c72a7d645fdf9eae24a5dd2fffa46685a5

jgowdyelastic commented on src/plugins/discover/public/application/main/components/pattern_analysis/pattern_analysis_table.tsx:
```diff
@@ -0,0 +1,79 @@
+
+  useEffect(() => {
+    // Track should only be called once when component is loaded
+    trackUiMetric?.(METRIC_TYPE.LOADED, PATTERN_ANALYSIS_LOADED);
```
Updated in https://github.com/elastic/kibana/pull/178916/commits/299621c72a7d645fdf9eae24a5dd2fffa46685a5

jgowdyelastic commented on src/plugins/discover/public/application/main/components/pattern_analysis/pattern_analysis_table.tsx:
```diff
@@ -0,0 +1,79 @@
+        input,
+        renderViewModeToggle: props.renderViewModeToggle,
+      }}
+      deps={deps}
```
Updated in https://github.com/elastic/kibana/pull/178916/commits/299621c72a7d645fdf9eae24a5dd2fffa46685a5

jgowdyelastic commented on src/plugins/discover/public/application/main/components/pattern_analysis/pattern_analysis_table.tsx:
```diff
@@ -0,0 +1,79 @@
+
+  return (
+    <aiopsService.PatternAnalysisComponent
+      props={{
```
Updated in https://github.com/elastic/kibana/pull/178916/commits/299621c72a7d645fdf9eae24a5dd2fffa46685a5

jgowdyelastic commented on x-pack/plugins/aiops/public/components/log_categorization/category_table/category_table.tsx:
```diff
@@ -302,49 +249,49 @@ export const CategoryTable: FC<Props> = ({
-      />
-    </>
+  return (
+    <EuiInMemoryTable<Category>
```
Updated in https://github.com/elastic/kibana/pull/178916/commits/299621c72a7d645fdf9eae24a5dd2fffa46685a5

jgowdyelastic commented on x-pack/plugins/aiops/public/components/log_categorization/log_categorization_enabled.ts:
```diff
@@ -0,0 +1,20 @@
+  return async (dataView: DataView) => {
+    const hasTextFields = dataView.fields.some((f) => f.esTypes?.includes(ES_FIELD_TYPES.TEXT));
+    const isPlatinum = (await lic).hasAtLeast('platinum');
+    return isPlatinum && hasTextFields && dataView.isTimeBased();
```
Updated in https://github.com/elastic/kibana/pull/178916/commits/299621c72a7d645fdf9eae24a5dd2fffa46685a5

jgowdyelastic commented on x-pack/plugins/aiops/public/components/log_categorization/log_categorization_for_embeddable/discover_tabs.tsx:
```diff
@@ -0,0 +1,115 @@
+                      categoryCount={data?.totalCategories}
+                    />
+                  </div>
+                  <div className="unifiedDataTableToolbarControlIconButton">
```
Updated in https://github.com/elastic/kibana/pull/178916/commits/299621c72a7d645fdf9eae24a5dd2fffa46685a5

jgowdyelastic commented on x-pack/plugins/aiops/public/components/log_categorization/log_categorization_for_embeddable/discover_tabs.tsx:
```diff
@@ -0,0 +1,115 @@
+              <EuiFlexItem>
+                <div
+                  className="unifiedDataTableToolbarControlGroup"
+                  color="subdued"
```
Updated in https://github.com/elastic/kibana/pull/178916/commits/299621c72a7d645fdf9eae24a5dd2fffa46685a5

qn895 commented:
Tested latest changes and ML-changes LGTM. Please feel free to merge this one ahead of [the ES|QL field stats in Discover PR](https://github.com/elastic/kibana/pull/180849) 👍 . I'll wait to merge it after this one is in.

qn895 commented on src/plugins/discover/public/application/main/components/pattern_analysis/pattern_analysis_table.tsx:
```diff
@@ -0,0 +1,71 @@
+import { METRIC_TYPE, UiCounterMetricType } from '@kbn/analytics';
+import { type EmbeddablePatternAnalysisInput } from '@kbn/aiops-log-pattern-analysis/embeddable';
+import { pick } from 'lodash';
+import { LogCategorizationEmbeddableProps } from '@kbn/aiops-plugin/public/components/log_categorization/log_categorization_for_embeddable/log_categorization_for_embeddable';
```
nit: import type

jgowdyelastic replied:
Updated in https://github.com/elastic/kibana/pull/178916/commits/1aaa89fc3a28c496c06e84f5335fd50ae31732d2

davismcphee commented on src/plugins/saved_search/server/saved_objects/search.ts:
```diff
@@ -62,6 +63,13 @@ export function getSavedSearchObjectType(
         },
       },
+      4: {
+        changes: [],
```
I don't think a `change` is necessary here since it's only a validation schema change. These are the current supported change types and they're all based around mapping changes or data modification which isn't needed here: https://github.com/elastic/kibana/blob/e2ccaf2c5305b7c3843d77a7aed20306eb0bf73f/packages/core/saved-objects/core-saved-objects-server/src/model_version/model_change.ts#L30-L35

andreadelrio commented:
Hi @jgowdyelastic sending design comments below:

- Add a heading to the `Pattern analysis settings` popover. For reference, this should look like the `Keyboard shortcuts` popover in the `Document` tab.

<img width="952" alt="Frame 1" src="https://github.com/elastic/kibana/assets/4016496/ca9d6dac-c4e7-4260-8b9f-62c9c03b0ee6">

- Increase the padding on the left side of the Patterns tab to achieve something similar to what we have in the `Documents` tab.

<img width="923" alt="Frame 9" src="https://github.com/elastic/kibana/assets/4016496/dae3342b-7a7d-44ac-bfbd-16f698cdf2b6">

**Long term** and as I mentioned to Aris, it would be beneficial to consider using **EuiDataGrid** for this view as well. It would match nicely with the `Documents` tab. On that note, given the amount of information you're displaying when expanding the row, I would suggest considering moving it to a flyout. This would be similar to how in `Documents` tab the user presses the `expand` icon and the push flyout appears. The pattern shown in the flyout would be highlighted in the main Patterns table. Here's a **quick** mockup of how that flyout could look like:

<img width="489" alt="Flyout" src="https://github.com/elastic/kibana/assets/4016496/3cb184a6-005a-4725-8d7c-6da8b052a8a1">

cc @kertal @davismcphee @ryankeairns for awareness.

jgowdyelastic commented on src/plugins/discover/public/application/main/components/pattern_analysis/pattern_analysis_tab.tsx:
```diff
@@ -0,0 +1,36 @@
+        dataView={props.dataView}
+        filters={querySubscriberResult.filters}
+        query={querySubscriberResult.query}
+        onAddFilter={props.onAddFilter}
```
I've renamed this prop to `switchToDocumentView` as this is all it is indented to do.

jgowdyelastic commented on src/plugins/discover/public/components/view_mode_toggle/view_mode_toggle.tsx:
```diff
@@ -84,16 +134,33 @@ export const DocumentViewModeToggle = ({
+              >
+                <FormattedMessage
+                  id="discover.viewModes.patternAnalysis.label"
+                  defaultMessage="Patterns {patternCount}"
```
IMO it is necessary to show the count and having it in the tab matches the document count, but I agree that it's not ideal that it shifts the field stats over when it renders.
The simplest solution would be to move the Patterns tab over so it is the 3rd tab. It was originally requested that it should be the second tab, I believe for visibility, but I can't imagine it makes much difference. 
What do you think @arisonl ?

arisonl replied:
Agreed 100% with @jgowdyelastic. The count is very important (as users filter things for and out based on that count), the relative tab placement less so.

arisonl replied:
++ the count is absolutely necessary as @jgowdyelastic mentions (it is the measure via which a user would choose to interact with the feature, i.e.filter patterns for or out).

I believe that it is also important that Patterns is the second tab: First, I believe that the third tab will get less engagement than the second. We know from data that currently the second tab is getting a 30% at maximum and Patterns is more important than stats, we would not want to start it lower than that. Second, Patterns and Documents work together (you filter in the one and you land to the other), so in terms of user interactions and the UX's information structure, field statistics is the "odd" one in this triplet and ideally it should not stand in between.

jgowdyelastic commented:
@andreadelrio 

> 
> * Add a heading to the `Pattern analysis settings` popover. For reference, this should look like the `Keyboard shortcuts` popover in the `Document` tab.
> 

Updated in https://github.com/elastic/kibana/pull/178916/commits/13e4b30ad7340ca8143802aa9c735755c4f65280

> * Increase the padding on the left side of the Patterns tab to achieve something similar to what we have in the `Documents` tab.

Updated in. https://github.com/elastic/kibana/pull/178916/commits/13e4b30ad7340ca8143802aa9c735755c4f65280

> **Long term** and as I mentioned to Aris, it would be beneficial to consider using **EuiDataGrid** for this view as well. It would match nicely with the `Documents` tab. On that note, given the amount of information you're displaying when expanding the row, I would suggest considering moving it to a flyout. This would be similar to how in `Documents` tab the user presses the `expand` icon and the push flyout appears. The pattern shown in the flyout would be highlighted in the main Patterns table. Here's a **quick** mockup of how that flyout could look like:

Yeah, I think this would be an improvement to the expanded rows.
I think we would also need to navigation controls similar to the document flyout to allow users to easily move between patterns. e.g.
![image](https://github.com/elastic/kibana/assets/22172091/95335e7a-b5a3-4584-9b09-e3596aaff5e7)
I can look into this for a follow up PR.

peteharverson reviewed (approved):
Tested latest changes and LGTM.

jughosta reviewed (approved):
Great feature! And thanks for addressing the comments!

Some things to consider improving as a follow up:

-  What are these warnings `0% contain 3 or more tokens` about? The shown patterns seem to be fine:

<img width="899" alt="Screenshot 2024-05-22 at 10 52 20" src="https://github.com/elastic/kibana/assets/1415710/02d02054-42cf-4eba-b277-34558afa2e9b">

- Noticed that cancelling a patterns request does not always work right away. Is it expected? 

- Once canceled, should there be a button to restart? 

<img width="969" alt="Screenshot 2024-05-22 at 10 41 07" src="https://github.com/elastic/kibana/assets/1415710/0d02709f-c9f7-4ed7-84f4-8016ed38eaec">

Style nits:

- The label is to close to the side:

<img width="220" alt="Screenshot 2024-05-22 at 10 39 58" src="https://github.com/elastic/kibana/assets/1415710/a717ccef-7b80-4988-aa1e-abd37fe2b1a5">

- Mobile view is a little off:

<img width="542" alt="Screenshot 2024-05-22 at 10 37 38" src="https://github.com/elastic/kibana/assets/1415710/355b7252-1da2-41dd-b96f-658f6c0907fc">

<img width="534" alt="Screenshot 2024-05-22 at 10 57 08" src="https://github.com/elastic/kibana/assets/1415710/926a990c-c81d-443d-98ec-39763239de71">

<img width="937" alt="Screenshot 2024-05-22 at 10 56 28" src="https://github.com/elastic/kibana/assets/1415710/7d099e2b-94e0-40fd-a37e-e382e709dcb4">

jgowdyelastic commented:
Follow on tasks have been added to this meta issue https://github.com/elastic/kibana/issues/183922
